builtin/logical/transit/policy.go

```diff
@@ -72,11 +73,143 @@ type Policy struct {
+	// Encode the policy
+	buf, err := json.Marshal(archive)
+	if err != nil {
+		return err
+	}
+
+	// Write the policy into storage
+	err = storage.Put(&logical.StorageEntry{
+		Key:   "policy/" + p.Name + "/archive",
+		Value: buf,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// handleArchiving manages the movement of keys to and from the policy archive.
+// This should *ONLY* be called from Persist() since it assumes that the policy
+// will be persisted afterwards.
+func (p *Policy) handleArchiving(storage logical.Storage) error {
+	// We need to move keys that are no longer accessible to ArchivedKeys, and keys
+	// that now need to be accessible back here.
+	//
+	// For safety, because there isn't really a good reason to, we never delete
+	// keys from the archive even when we move them back.
+
+	// Check if we have the latest minimum version in the current set of keys
+	_, keysContainsMinimum := p.Keys[p.MinDecryptionVersion]
+
+	// Sanity checks
+	switch {
+	case p.MinDecryptionVersion < 1:
+		return fmt.Errorf("minimum decryption version of %d is less than 1", p.MinDecryptionVersion)
+	case p.LatestVersion < 1:
+		return fmt.Errorf("latest version of %d is less than 1", p.LatestVersion)
+	case !keysContainsMinimum && p.ArchiveVersion != p.LatestVersion:
+		return fmt.Errorf("need to move keys from archive but archive version not up-to-date")
+	case p.ArchiveVersion > p.LatestVersion:
+		return fmt.Errorf("archive version of %d is greater than the latest version %d",
+			p.ArchiveVersion, p.LatestVersion)
+	case p.MinDecryptionVersion > p.LatestVersion:
+		return fmt.Errorf("minimum decryption version of %d is greater than the latest version %d",
+			p.MinDecryptionVersion, p.LatestVersion)
+	}
+
+	archive, err := p.loadArchive(storage)
+	if err != nil {
+		return err
+	}
+
+	if keysContainsMinimum {
+		// Need to move keys *to* archive
+
+		// We need a size that is equivalent to the latest version (number of
+		// keys) but adding one since slice numbering starts at 0 and we're
+		// indexing by key version
+		if len(archive.Keys) < p.LatestVersion+1 {
+			// Increase the size of the archive slice
+			newKeys := make([]KeyEntry, p.LatestVersion+1)
+			copy(newKeys, archive.Keys)
+			archive.Keys = newKeys
+		}
+
+		// We are storing all keys in the archive, so we ensure that it is up
+		// to date up to p.LatestVersion
+		for i := p.ArchiveVersion + 1; i <= p.LatestVersion; i++ {
+			archive.Keys[i] = p.Keys[i]
+			p.ArchiveVersion = i
+		}
+
+		err = p.storeArchive(archive, storage)
+		if err != nil {
+			return err
+		}
+
+		// Perform deletion afterwards so that if there is an error saving we
+		// haven't messed with the current policy
+		for i := p.LatestVersion - len(p.Keys) + 1; i < p.MinDecryptionVersion; i++ {
+			delete(p.Keys, i)
+		}
+
+	} else {
+		// Need to move keys *from* archive
+
+		for i := p.MinDecryptionVersion; i <= p.LatestVersion; i++ {
```

Comment:
Can we just move this into a `if ! keysContainsMinimum` block and then do an early exist. This way we can reduce the nest level of the above block